### DISCUSSION TITLE
Check real time operation with doMPC

### DISCUSSION THREAD

**Discussion by wissamkafa** [do-mpc/do-mpc#386] [Q&A]
Hi
when applying a simulation model using doMPC on PC, can I check if this controller will work in real time when implementing it in boards like jetson, arduino, or raspberry?

I executed the model on two PCs and each of them gives different execution time (which is normal because of different specifications).
but for example, if I need a controller to work in response time like 0.1s, can I check if the code and model I am implementing will work on real time before implementing it on real system?
is it like that:
if the optimizer shows "EXIT: Optimal Solution Found. average: 256.06ms" on core i7 , 2.7GH, PC, does this mean that I need higher competition power to make my system run in real time? and if so, what can I do to reduce the computation time?

**Comment by 4flixt** [comment]
Hi @wissamkafa,

there is no easy answer to this question. There are many things you could consider to improve the speed of do-mpc (apart from hardware improvements). You can modify your problem formulation (cost, model, constraints), use scaling and other tricks. You can also change the linear solver, e.g to MA27. 

I wouldn't be aware of a good method to predict the computation time on e.g. a raspberry pi without testing it. 

Best
4flixt

**Reply by wissamkafa** [reply]
**Thanks for your reply.**
I agree with that, I am already trying to change the mumps solver to MA27, but with no success.

_according to the instruction in do-MPC documentation
"
_**To use MA27** (or other HSL solver in this setup) please follow these steps:
Linux
(Tested on Ubuntu 19.10)
    1- Obtain the [HSL](http://www.hsl.rl.ac.uk/ipopt/) shared library. Choose the personal licence.
    2- Unpack the archive and copy its content to a destination of your choice. (e.g. /home/username/Documents/coinhsl/)
    3- Rename libcoinhsl.so to libhsl.so. CasADi is searching for the shared libraries under a depreciated name.
    4- Locate your .bashrc file on your home directory (e.g. /home/username/.bashrc)
    5- Add _the p_reviously created directory to your LD_LIBRARY_PATH, by adding the following line to your .bashrc_
"_
when we get the SHL archive we need to change **"libcoinhsl.so"** to "libhsl.so", but **there is no such file within the archive**. 
I could not figure a solution for this.
I also tried to follow the instruction in CasADi web site: https://github.com/casadi/casadi/wiki/Obtaining-HSL
but it does not work and many steps is confusing to me.

if could you please suggest a solution on how to adress the file "libcoinhsl.so", so i can continue the procedure mentioned in: 
https://www.do-mpc.com/en/v4.1.1/installation.html
I am using ubuntu 22.04 - python 3.10 - Anaconda3, pycharm editor
Best regards

**Comment by 4flixt** [comment]
HI @wissamkafa,

I think, unfortunately, HSL does not offer the compiled binaries of MA27 anymore? So you need to compile the c-code yourself on your machine, which should yield the libcoinhsl.so file (on Linux).
The process is tedious and I also struggled a bit when I last did it...
Maybe you find someone with c-code experience and knows how to compile to help you?

Best,
4flixt

**Reply by wissamkafa** [reply]
Hi,
Thank you for your help and your time, I appreciate this very much.

I tried to compile the HSL files using the commands from terminal as mentioned in (https://github.com/coin-or-tools/ThirdParty-HSL):

_1. ./configure
2. make
3. make install_

All runs without problems, and the ./configure odes detect that ma27 esist into the backage.
but libcoinhsl.so, has not been created into the directory i am using (where the extracted files exist), after long search i noticed that it has been created in another directory "usr/local/lib".
If i tried to rename it and move it to my directory, something goes wrong and does not work (an x sign appears as icon on it, and its size is not correct).
I included the command 
_export LD_LIBRARY_PATH="$LD_LIBRARY_PATH:/home/user/ThirdParty-HSL/coinhsl/lib"_
in the ~/.bashrc file
I keeps getting the error:
_Exception of type: DYNAMIC_LIBRARY_FAILURE in file "/work/build/external_projects/src/ipopt-external/src/Common/IpLibraryLoader.cpp" at line 67:
 Exception message: libhsl.so: cannot open shared object file: No such file or directory_

which means that step 3 in your instructions does not exicuted correctly.

Then I tried to leave the libcoinhsl.so file where it was generated (usr/local/lib) and also included this path into the .bashrc file,
also I got the same error.

Note:
I am using this code in template_mpc:
_setup_mpc = {
        'n_robust': 1,
        'n_horizon': 10,#20
        't_step': 0.1,
        'store_full_solution': True,
        # Use MA27 linear solver in ipopt for faster calculations:
        'nlpsol_opts': {'ipopt.linear_solver': 'MA27'},
    }_
    
do you think I am missing somthing here? 
actually I tried alot and do not what options left
